src/primitives/polynomial.rs

```diff
@@ -132,75 +135,205 @@ impl<F: Field> Polynomial<F> {
+        // of the group generated by prim_elem. In *theory* this means that we
         // should go back and find the second-longest geometric series and try
         // that, because for a real-life BCH code this situation indicates that
         // something is wrong and we should just panic.
-        let code_len = max_ratio.multiplicative_order();
-
+        let initial_elem = E::GENERATOR.powi(max_start as i64);
         let mut min_index = None;
-        let mut base = F::ONE;
+        let mut base = E::ONE;
         for i in 0..code_len {
-            base *= &max_ratio;
-            if base == max_start {
+            if base == initial_elem {
                 min_index = Some(i);
             }
+            base *= &prim_elem;
         }
-
         let min_index = match min_index {
             Some(idx) => idx,
-            None => panic!("Found geometric series within roots starting from {} (ratio {} length {}), but this series does not consist of powers of any generator.", max_start, max_ratio, max_length),
+            None => panic!("Found geometric series within roots starting from {} (ratio {} length {}), but the series does not consist of powers of the ratio.", initial_elem, prim_elem, max_length),
         };
 
         // We write `a..=b - 1` instead of `a..b` because RangeInclusive is actually
         // a different type than Range, so the two syntaxes are not equivalent here.
-        (max_ratio, code_len, min_index..=min_index + max_length - 1)
+        (prim_elem, code_len, min_index..=min_index + max_length - 1)
+    }
+}
+
+impl<F: Field> fmt::Display for Polynomial<F> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.has_data() {
+            for fe in self.iter() {
+                write!(f, "{}", fe)?;
+            }
+            Ok(())
+        } else {
+            f.write_str("<residue>")
+        }
+    }
+}
+
+impl<F: Field> iter::FromIterator<F> for Polynomial<F> {
+    #[inline]
+    fn from_iter<I>(iter: I) -> Self
+    where
+        I: IntoIterator<Item = F>,
+    {
+        Polynomial { inner: FieldVec::from_iter(iter) }
+    }
+}
+
+impl<F> From<FieldVec<F>> for Polynomial<F> {
+    fn from(inner: FieldVec<F>) -> Self { Self { inner } }
+}
+
+impl<F: Field> ops::Add<&Polynomial<F>> for Polynomial<F> {
+    type Output = Polynomial<F>;
+
+    /// Interprets two residues as polynomials and adds them together.
+    ///
+    /// # Panics
+    ///
+    /// Will panic if [`Polynomial::has_data`] is false.
+    fn add(mut self, other: &Polynomial<F>) -> Polynomial<F> {
+        self += other;
+        self
+    }
+}
+
+impl<F: Field> ops::Add<Polynomial<F>> for Polynomial<F> {
+    type Output = Polynomial<F>;
+    fn add(self, other: Polynomial<F>) -> Polynomial<F> { self + &other }
+}
+
+impl<F: Field> ops::Sub<&Polynomial<F>> for Polynomial<F> {
+    type Output = Polynomial<F>;
+    fn sub(self, other: &Polynomial<F>) -> Polynomial<F> { self + other }
```

Comment:
This is finite field math so add and sub are the same, right? Perhaps add a comment for less cryptographically trained devs.

Reply:
In this case I probably should actually pass through to the sub impl for `F` (which is the same as the add impl in all cases, because I have a binary field, but this stuff should all still work even with non-binary fields).